[azdatalake] file.Client.GetProperties seg faults when the file exists

### Bug Report

Module version: `github.com/Azure/azure-sdk-for-go/sdk/storage/azdatalake v1.4.2`
Package: `github.com/Azure/azure-sdk-for-go/sdk/storage/azdatalake/file`
go version: `go version go1.24.0 linux/amd64`

When using `GetProperties` to determine if a file exists, I get a segfault from within the librari if the file exists. It works if the file does not.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0xf9c6d5]

goroutine 68 [running]:
github.com/Azure/azure-sdk-for-go/sdk/storage/azdatalake/internal/path.FormatGetPropertiesResponse(_, _)
        /home/jakedern/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/storage/azdatalake@v1.4.2/internal/path/responses.go:290 +0x315
github.com/Azure/azure-sdk-for-go/sdk/storage/azdatalake/file.(*Client).GetProperties(_, {_, _}, _)
        /home/jakedern/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/storage/azdatalake@v1.4.2/file/client.go:241 +0x1cd
dev.azure.com/msdata/Tina/monitoring/pkg/objectstore.(*dataLakeStorage).Read(0xc000313f00, {0xc000538000, 0x23}, 0xc000536300)
        /home/jakedern/repos/monitoring/projects/go/pkg/objectstore/datalake.go:135 +0x208
dev.azure.com/msdata/Tina/monitoring/cmd/otelcol-arcdata/internal/exporter/fabricexporter.(*fabricExporter[...]).initializeTable(0x3cb6820, {0x3cd2370, 0x5ed37a0}, 0x19)
        /home/jakedern/repos/monitoring/projects/go/cmd/otelcol-arcdata/internal/exporter/fabricexporter/exporter.go:311 +0x293
dev.azure.com/msdata/Tina/monitoring/cmd/otelcol-arcdata/internal/exporter/fabricexporter.(*fabricExporter[...]).start.func1()
        /home/jakedern/repos/monitoring/projects/go/cmd/otelcol-arcdata/internal/exporter/fabricexporter/exporter.go:163 +0x33
dev.azure.com/msdata/Tina/monitoring/cmd/otelcol-arcdata/internal/exporter/fabricexporter.runManyInParallel.func1(0x0?)
        /home/jakedern/repos/monitoring/projects/go/cmd/otelcol-arcdata/internal/exporter/fabricexporter/exporter.go:386 +0x53
created by dev.azure.com/msdata/Tina/monitoring/cmd/otelcol-arcdata/internal/exporter/fabricexporter.runManyInParallel in goroutine 1
        /home/jakedern/repos/monitoring/projects/go/cmd/otelcol-arcdata/internal/exporter/fabricexporter/exporter.go:384 +0x6b
```

--- a/sdk/storage/azdatalake/CHANGELOG.md
+++ b/sdk/storage/azdatalake/CHANGELOG.md
@@ -7,6 +7,7 @@
 ### Breaking Changes
 
 ### Bugs Fixed
+* Fixed a panic in `GetProperties` (and `DownloadStream`) for file and directory clients created via `directory.NewFileClient`/`directory.NewSubdirectoryClient`. These clients did not capture the raw HTTP response, so reading the datalake-specific headers dereferenced a nil response. Fixes [#25490](https://github.com/Azure/azure-sdk-for-go/issues/25490).
 
 ### Other Changes
 * Updated code generator to `@autorest/go@4.0.0-preview.80`.

--- a/sdk/storage/azdatalake/directory/client.go
+++ b/sdk/storage/azdatalake/directory/client.go
@@ -212,6 +212,10 @@ func (d *Client) NewFileClient(fileName string) (*file.Client, error) {
 	newBlobURL, fileURL := shared.GetURLs(fileURL)
 	var newBlobClient *blockblob.Client
 	clientOptions := &blockblob.ClientOptions{ClientOptions: d.getClientOptions().ClientOptions}
+	// The blob client built here must capture the raw response so that file.GetProperties
+	// can read the datalake-specific headers (owner, group, permissions). Without this the
+	// captured response is nil and GetProperties panics. See issue #25490.
+	clientOptions.PerCallPolicies = append([]policy.Policy{shared.NewIncludeBlobResponsePolicy()}, clientOptions.PerCallPolicies...)
 	var err error
 	if d.identityCredential() != nil {
 		newBlobClient, err = blockblob.NewClient(newBlobURL, *d.identityCredential(), clientOptions)
@@ -235,6 +239,10 @@ func (d *Client) NewSubdirectoryClient(subdirectoryName string) (*Client, error)
 	newBlobURL, subDirectoryURL := shared.GetURLs(subDirectoryURL)
 	var newBlobClient *blockblob.Client
 	clientOptions := &blockblob.ClientOptions{ClientOptions: d.getClientOptions().ClientOptions}
+	// The blob client built here must capture the raw response so that GetProperties can read
+	// the datalake-specific headers (owner, group, permissions). Without this the captured
+	// response is nil and GetProperties panics. See issue #25490.
+	clientOptions.PerCallPolicies = append([]policy.Policy{shared.NewIncludeBlobResponsePolicy()}, clientOptions.PerCallPolicies...)
 	var err error
 	if d.identityCredential() != nil {
 		newBlobClient, err = blockblob.NewClient(newBlobURL, *d.identityCredential(), clientOptions)

--- a/sdk/storage/azdatalake/file/responses.go
+++ b/sdk/storage/azdatalake/file/responses.go
@@ -242,11 +242,15 @@ func FormatDownloadStreamResponse(r *blob.DownloadStreamResponse, rawResponse *h
 		newResp.Version = r.Version
 		newResp.VersionID = r.VersionID
 	}
-	if val := rawResponse.Header.Get("x-ms-encryption-context"); val != "" {
-		newResp.EncryptionContext = &val
-	}
-	if val := rawResponse.Header.Get("x-ms-acl"); val != "" {
-		newResp.AccessControlList = &val
+	// rawResponse is only populated when the IncludeBlobResponsePolicy captured it.
+	// Guard against nil so the datalake-specific headers below never panic.
+	if rawResponse != nil {
+		if val := rawResponse.Header.Get("x-ms-encryption-context"); val != "" {
+			newResp.EncryptionContext = &val
+		}
+		if val := rawResponse.Header.Get("x-ms-acl"); val != "" {
+			newResp.AccessControlList = &val
+		}
 	}
 	return newResp
 }

--- a/sdk/storage/azdatalake/file/responses_test.go
+++ b/sdk/storage/azdatalake/file/responses_test.go
@@ -1,0 +1,37 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+package file
+
+import (
+	"net/http"
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/blob"
+	"github.com/stretchr/testify/require"
+)
+
+// FormatDownloadStreamResponse must not panic when the raw HTTP response was not
+// captured (rawResponse is nil). This mirrors the guard tested for
+// FormatGetPropertiesResponse. See issue #25490.
+func TestFormatDownloadStreamResponseNilRawResponse(t *testing.T) {
+	require.NotPanics(t, func() {
+		resp := FormatDownloadStreamResponse(&blob.DownloadStreamResponse{}, nil)
+		// The datalake-specific fields come from the raw response, so they stay nil.
+		require.Nil(t, resp.AccessControlList)
+		require.Nil(t, resp.EncryptionContext)
+	})
+}
+
+// When the raw HTTP response is present, the datalake-specific headers are read.
+func TestFormatDownloadStreamResponseWithRawResponse(t *testing.T) {
+	rawResponse := &http.Response{Header: http.Header{}}
+	rawResponse.Header.Set("x-ms-acl", "user::rwx,group::r-x,other::---")
+	rawResponse.Header.Set("x-ms-encryption-context", "context1")
+
+	resp := FormatDownloadStreamResponse(&blob.DownloadStreamResponse{}, rawResponse)
+	require.NotNil(t, resp.AccessControlList)
+	require.Equal(t, "user::rwx,group::r-x,other::---", *resp.AccessControlList)
+	require.NotNil(t, resp.EncryptionContext)
+	require.Equal(t, "context1", *resp.EncryptionContext)
+}

--- a/sdk/storage/azdatalake/internal/path/responses.go
+++ b/sdk/storage/azdatalake/internal/path/responses.go
@@ -287,23 +287,27 @@ func FormatGetPropertiesResponse(r *blob.GetPropertiesResponse, rawResponse *htt
 	newResp.TagCount = r.TagCount
 	newResp.Version = r.Version
 	newResp.VersionID = r.VersionID
-	if val := rawResponse.Header.Get("x-ms-owner"); val != "" {
-		newResp.Owner = &val
-	}
-	if val := rawResponse.Header.Get("x-ms-group"); val != "" {
-		newResp.Group = &val
-	}
-	if val := rawResponse.Header.Get("x-ms-permissions"); val != "" {
-		newResp.Permissions = &val
-	}
-	if val := rawResponse.Header.Get("x-ms-acl"); val != "" {
-		newResp.AccessControlList = &val
-	}
-	if val := rawResponse.Header.Get("x-ms-resource-type"); val != "" {
-		newResp.ResourceType = &val
-	}
-	if val := rawResponse.Header.Get("x-ms-encryption-context"); val != "" {
-		newResp.EncryptionContext = &val
+	// rawResponse is only populated when the IncludeBlobResponsePolicy captured it.
+	// Guard against nil so the datalake-specific headers below never panic.
+	if rawResponse != nil {
+		if val := rawResponse.Header.Get("x-ms-owner"); val != "" {
+			newResp.Owner = &val
+		}
+		if val := rawResponse.Header.Get("x-ms-group"); val != "" {
+			newResp.Group = &val
+		}
+		if val := rawResponse.Header.Get("x-ms-permissions"); val != "" {
+			newResp.Permissions = &val
+		}
+		if val := rawResponse.Header.Get("x-ms-acl"); val != "" {
+			newResp.AccessControlList = &val
+		}
+		if val := rawResponse.Header.Get("x-ms-resource-type"); val != "" {
+			newResp.ResourceType = &val
+		}
+		if val := rawResponse.Header.Get("x-ms-encryption-context"); val != "" {
+			newResp.EncryptionContext = &val
+		}
 	}
 	return newResp
 }

--- a/sdk/storage/azdatalake/internal/path/responses_test.go
+++ b/sdk/storage/azdatalake/internal/path/responses_test.go
@@ -1,0 +1,42 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+package path
+
+import (
+	"net/http"
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/blob"
+	"github.com/stretchr/testify/require"
+)
+
+// FormatGetPropertiesResponse must not panic when the raw HTTP response was not
+// captured (rawResponse is nil). This happens for clients whose pipeline does not
+// include the IncludeBlobResponsePolicy. See issue #25490.
+func TestFormatGetPropertiesResponseNilRawResponse(t *testing.T) {
+	require.NotPanics(t, func() {
+		resp := FormatGetPropertiesResponse(&blob.GetPropertiesResponse{}, nil)
+		// The datalake-specific fields come from the raw response, so they stay nil.
+		require.Nil(t, resp.Owner)
+		require.Nil(t, resp.Group)
+		require.Nil(t, resp.Permissions)
+		require.Nil(t, resp.AccessControlList)
+		require.Nil(t, resp.ResourceType)
+	})
+}
+
+// When the raw HTTP response is present, the datalake-specific headers are read.
+func TestFormatGetPropertiesResponseWithRawResponse(t *testing.T) {
+	rawResponse := &http.Response{Header: http.Header{}}
+	rawResponse.Header.Set("x-ms-owner", "owner1")
+	rawResponse.Header.Set("x-ms-group", "group1")
+	rawResponse.Header.Set("x-ms-permissions", "rwxr-x---")
+	rawResponse.Header.Set("x-ms-resource-type", "file")
+
+	resp := FormatGetPropertiesResponse(&blob.GetPropertiesResponse{}, rawResponse)
+	require.Equal(t, "owner1", *resp.Owner)
+	require.Equal(t, "group1", *resp.Group)
+	require.Equal(t, "rwxr-x---", *resp.Permissions)
+	require.Equal(t, "file", *resp.ResourceType)
+}

